okapi-common/src/main/java/org/folio/okapi/common/refreshtoken/client/impl/LoginClient.java

```diff
@@ -0,0 +1,145 @@
+package org.folio.okapi.common.refreshtoken.client.impl;
+
+import io.netty.handler.codec.http.cookie.ClientCookieDecoder;
+import io.netty.handler.codec.http.cookie.Cookie;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.client.HttpRequest;
+import java.util.function.Supplier;
+import org.folio.okapi.common.Constants;
+import org.folio.okapi.common.XOkapiHeaders;
+import org.folio.okapi.common.refreshtoken.client.Client;
+import org.folio.okapi.common.refreshtoken.client.ClientException;
+import org.folio.okapi.common.refreshtoken.client.ClientOptions;
+import org.folio.okapi.common.refreshtoken.tokencache.TenantUserCache;
+
+@java.lang.SuppressWarnings({"squid:S1075"}) // URIs should not be hardcoded
+public class LoginClient implements Client {
+
+  private static final String LOGIN_EXPIRY_PATH = "/authn/login-with-expiry";
+
+  private static final String LOGIN_LEGACY_PATH = "/authn/login";
+
+  private final TenantUserCache cache;
+
+  private final ClientOptions clientOptions;
+
+  private final String tenant;
+
+  private final String username;
+
+  private final Supplier<Future<String>> getPasswordSupplier;
+
+  /**
+   * Refresh legacy tokens older than this.
+   */
+  private static final long AGE_LEGACY_TOKEN = 86400L;
+
+  /**
+   * Subtract this many seconds of age, before considering expired.
+   */
+  private static final long AGE_DIFF_TOKEN = 10L;
```

Comment:
We can never be sure. I don't mind using what you suggest which is  more conservative.

The token is checked by mod-authtoken in the first stages of request handling of Okapi. Even so, if the client code (using this API), is running out of sockets in the pool it may take time before the request is sent.

Reply:
For requests to mod-authtoken which is the first thing that happens. mod-authtoken makes a new "dummy" token at this stage.

Changed to age/2 as this most likely is more conservative.